sslr-core/src/main/java/com/sonar/sslr/impl/typed/SyntaxTreeCreator.java

```diff
@@ -0,0 +1,191 @@
+import com.sonar.sslr.api.GenericTokenType;
+import com.sonar.sslr.api.Token;
+import com.sonar.sslr.api.TokenType;
+import com.sonar.sslr.api.Trivia;
+import com.sonar.sslr.api.Trivia.TriviaKind;
+import com.sonar.sslr.api.typed.ActionParser.GrammarBuilderInterceptor;
+import com.sonar.sslr.api.typed.Input;
+import com.sonar.sslr.api.typed.NodeBuilder;
+import com.sonar.sslr.api.typed.Optional;
+import org.sonar.sslr.grammar.GrammarRuleKey;
+import org.sonar.sslr.internal.grammar.MutableParsingRule;
+import org.sonar.sslr.internal.matchers.ParseNode;
+import org.sonar.sslr.internal.vm.TokenExpression;
+import org.sonar.sslr.internal.vm.TriviaExpression;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.List;
+
+public class SyntaxTreeCreator<T> {
+
+  private final Object treeFactory;
+  private final GrammarBuilderInterceptor mapping;
+  private final NodeBuilder nodeBuilder;
+
+  private final Token.Builder tokenBuilder = Token.builder();
+  private final List<Trivia> trivias = Lists.newArrayList();
+
+  private Input input;
+
+  public SyntaxTreeCreator(Object treeFactory, GrammarBuilderInterceptor mapping, NodeBuilder nodeBuilder) {
+    this.treeFactory = treeFactory;
+    this.mapping = mapping;
+    this.nodeBuilder = nodeBuilder;
+  }
+
+  public T create(ParseNode node, Input input) {
+    this.input = input;
+    this.trivias.clear();
+    T result = (T) visit(node);
+    if (result instanceof AstNode) {
+      ((AstNode) result).hasToBeSkippedFromAst();
+    }
+    return result;
+  }
+
+  private Object visit(ParseNode node) {
+    if (node.getMatcher() instanceof MutableParsingRule) {
+      return visitNonTerminal(node);
+    } else {
+      return visitTerminal(node);
+    }
+  }
+
+  private Object visitNonTerminal(ParseNode node) {
+    MutableParsingRule rule = (MutableParsingRule) node.getMatcher();
+    GrammarRuleKey ruleKey = rule.getRuleKey();
+
+    if (mapping.hasMethodForRuleKey(ruleKey)) {
+      // TODO Drop useless intermediate nodes
+      Preconditions.checkState(node.getChildren().size() == 1);
+      return visit(node.getChildren().get(0));
+    }
+
+    if (mapping.isOptionalRule(ruleKey)) {
+      Preconditions.checkState(node.getChildren().size() <= 1);
+      if (node.getChildren().isEmpty()) {
+        return Optional.absent();
+      } else {
+        Object child = visit(node.getChildren().get(0));
+        if (child instanceof AstNode) {
+          ((AstNode) child).hasToBeSkippedFromAst();
+        }
+        return Optional.of(child);
+      }
+    }
+
+    List<ParseNode> children = node.getChildren();
+    List<Object> convertedChildren = Lists.newArrayList();
+    for (ParseNode child : children) {
+      Object result = visit(child);
+
+      if (result != null) {
+        // FIXME to remove aafter full migration: Allow to skip optional nodes that are supposed to bw skipped from the AST
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)